### DISCUSSION TITLE
Applying texture from an image

### DISCUSSION THREAD

**Discussion by josepereiroml** [pyvista/pyvista#3297] [Q&A] [answered]
Hi I'm trying to apply this texture (not just visualizing it I want to corrugate the mesh based on the picture of the tire and mapping the texture).
![image](https://user-images.githubusercontent.com/65933374/189607731-d51fced0-83bc-4a7c-bb85-e5da079a1922.png)

To this mesh: 
[object.zip](https://github.com/pyvista/pyvista/files/9546213/object.zip)

The idea is that it looks like this: 
![image](https://user-images.githubusercontent.com/65933374/189608274-5909f3cf-dc2e-4586-ba36-889af621dbc3.png)

Does anyone know how to solve it? I'm trying to use texture mapping but it does not work.

**Comment by adeak** [comment, marked as answer]
To repeat an approach I suggested in slack:
1. define texture coordinates on your mesh (this is the hard part in general; your example mesh here is spherical enough that texture mapping to a sphere will work as a first guess)
2. compute scalars by interpolating into the texture using the texture coordinates
3. warp your mesh by these scalars.

Here's a messy minimal example using your data:
```py
import numpy as np
from scipy.interpolate import griddata
import pyvista as pv

mesh = pv.read('object.obj')
texture = pv.read_texture('texture.png')

# subdivide the mesh for better resolution later
mesh.subdivide(3, inplace=True)

# define texture coordinates: mesh is "sphere-like" enough
mesh.texture_map_to_sphere(inplace=True)

# look at the traditionally mapped texture
mesh.plot(texture=texture, screenshot='mapped.png')

# define an interpolator to look up pixel values by
# texture.to_array() seems broken as of 0.36, so work around that
unif = texture.to_image()  # UniformGrid holding the texture data
img = unif.point_data['PNGImage'][..., 0].reshape(unif.dimensions[:-1], order='F')/255  # assume greyscale uint8 with no transparency, convert to float
# input points are the pixel coordinates, input values are the pixel values,
# query points are the texture coordinates
i, j = np.mgrid[0:1:img.shape[0]*1j, 0:1:img.shape[1]*1j]
tex_coords = mesh.point_data['Texture Coordinates']
interpolated = griddata((i.ravel(), j.ravel()), img.ravel(), tuple(tex_coords.T))

# assign scalars and warp
mesh.point_data['depth'] = 1 - interpolated/10  # scalars in [0.9, 1] for warping
mesh.compute_normals(inplace=True)
mesh.warp_by_scalar(scalars='depth').plot(screenshot='warped.png')
```

The first plot is what texture mapping would give you, i.e. colouration of the surface mesh:
![mapped](https://user-images.githubusercontent.com/17914410/189667098-d2307148-fe16-4de5-8d13-9ddb3a353541.png)

Note that I subdivided your surface, because the more points you have, the more accurate warping will be at the end.

What we then do is
1. extract a `UniformGrid` containing the texture information (because `texture.to_array()` seems broken until https://github.com/pyvista/pyvista/pull/2729/ gets finalized and merged)
2. pull out its scalars (converted from uint8 values to floats in `[0, 1]`)
3. interpolate the texture coordinates using [`scipy.interpolate.griddata`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.griddata.html) (using the default linear interpolation)
4. define the warping scalars as `1 - interpolated/10`, i.e. where the original image was 255 we get 0.9, and where the original image was 0 we get 1. This means an 0.1 relief in the warping. You can play around with these numbers to get your desired behaviour.
5. compute normals for warping (we should fix `warp_by_scalar()` so that we don't have to do this manually)
6. warp by the scalars.

This is the result:
![warped](https://user-images.githubusercontent.com/17914410/189668021-95b176c8-ee4f-48a4-b750-412c29984880.png)

You can see that it matches the texture in the first plot. The result will depend a lot on the texture you use. It should be compatible with your mesh in the sense that the proportions aren't distorted too much after mapping. You might have to crop your image for instance, to remove the white stripes. And in any case mapping a rectangle to a sphere will look weird unless the rectangle was a spherical projection to begin with (think globe maps).

Two things you can play with even in this example:
1. you can swap the two texture coordinate components to have the tyre texture along longitude rather than latitude (just do something like `tex_coords = tex_coords[:, ::-1]` to swap the two coordinates)
2. you can blur the original texture to get a smoother transition from surface to groove (one way to do this is to call `unif.gaussian_smooth()` with potential optional parameters).

This is the result of swapping the texture coordinates:

![warped and transposed](https://user-images.githubusercontent.com/17914410/189668924-72b1f6ee-cb89-434d-bda1-8f4a879823b8.png)

And this is the result of smoothing the texture on top of that:

![warped and smoothed](https://user-images.githubusercontent.com/17914410/189669036-d32fc0e9-3bf3-42d5-9fbb-a0cfa8e590c7.png)

**Reply by josepereiroml** [reply]
Thank you!! This is amazing I was not using the grid data.

**Reply by josepereiroml** [reply]
Hey I've been trying to use mesh.texture_map_to_plane(inplace=True) instead of the mesh.texture_map_to_sphere(inplace=True) but I'm getting this output 
![image](https://user-images.githubusercontent.com/65933374/193154842-b0c45883-0795-47b3-852c-13c7483dcfdc.png)
It looks like it's cutting the mesh but I don't understand why

**Reply by adeak** [reply]
Your mesh also seems different here, so: can you post a complete example with this? I assume the code only changed in that one line you mentioned, but it's better to make sure.
